Enables SSL_SASL

abohmeed opened the pull request:
<!--
Thanks for sending a pull request! 
If this is your first time, please read our contributor guidelines:
https://docs.seldon.io/projects/seldon-core/en/latest/developer/contributing.html
-->

**What this PR does / why we need it**:
Enables SASL_SSL for Kafka cluster connections
**Which issue(s) this PR fixes**:
https://github.com/SeldonIO/seldon-core/issues/4415
<!--
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->
Fixes #

**Special notes for your reviewer**:

abohmeed commented:
/assign @rafalskolasinski

abohmeed commented:
@RafalSkolasinski I've reopened the PR since the last one was incorrect

RafalSkolasinski commented:
@abohmeed thanks for linking the issue - it may be better to re-open the previous PR as @agrski already left some review comments there? There is no need to re-opening as new PR.

abohmeed commented:
@RafalSkolasinski Thanks. The old PR had unnecessary files modified (like topics.go) on which @agrski placed comments. So, I had to fix my branch and reopen a new PR where only two files are changed: server.go and utils.go. Sorry for any inconvenience.

RafalSkolasinski commented:
@abohmeed Once you fix your branch and push the PR should automatically update itself. If this is not the case it could be because you opened it from your `master` branch - you could then try to create feature branch in your fork and open a PR from it. Other than changing the source branch of the PR you should not need to open a new one. Unless I am missing something?

abohmeed commented:
@RafalSkolasinski Can't we just use this PR instead? I'm afraid I cannot reopen the old one since I re-forked the repo to avoid any issues.

RafalSkolasinski commented:
Okay, re-forking probably changes the situation. Let's continue here then :+1:

abohmeed commented:
@agrski I've addressed your comments from the previous PR (closed) except for this one https://github.com/SeldonIO/seldon-core/pull/4414#discussion_r1009392524. Can we open a new conversation about it in this PR?

seldondev commented:
@abohmeed: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | 35deeab2aee0837527fbb2c7bd4f5cc3da772f0d | [link](http://dashboard-jx.34.121.152.94.nip.io/namespaces/jx/pipelineruns/seldon-core-pr-4416-integration-l5868) | `/test integration`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

RafalSkolasinski commented:
We have `executor` tests failing with
```
vet: api/util/utils.go:195:3: unknown field Password in struct literal
```

RafalSkolasinski commented:
@abohmeed you should be able to run these tests locally with `make test` inside `executor` directory.

abohmeed commented:
@RafalSkolasinski Thanks for getting back to me. 
I made the necessary fixes and ran `make test` locally and it passed.

RafalSkolasinski commented:
/test integration

agrski reviewed (commented):
Thanks again @abohmeed for improving the Kafka support here.  Thanks also for tidying up the PR to   
These comments aren't blockers, but it'd be good to discuss/implement them to make this PR even more useful.

On the other hand, I think this PR *should* also update `executor/logger/worker.go` to add in equivalent config handling there.  This is the component that writes inference logs to Kafka (as opposed to a REST endpoint).

abohmeed commented:
/test integration

seldondev commented:
@abohmeed: Cannot trigger testing until a trusted user reviews the PR and leaves an `/ok-to-test` message.

<details>

In response to [this](https://github.com/SeldonIO/seldon-core/pull/4416#issuecomment-1299939393):

>/test integration
>


Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository.
</details>

abohmeed commented:
@agrski Thanks for your comments. I've addressed both points.

abohmeed commented:
@RafalSkolasinski @agrski Any updates? 🙂

agrski commented:
> @RafalSkolasinski @agrski Any updates? slightly_smiling_face

Hi @abohmeed, just getting back to this now.  Thanks for being so responsive and applying the recommendations.

agrski commented on executor/api/kafka/server.go:
```diff
@@ -101,7 +101,22 @@ func NewKafkaServer(fullGraph bool, workers int, deploymentName, namespace, prot
+				}
+			}
+		}
+		if util.GetKafkaSecurityProtocol() == "SASL_PLAIN" { //if we also have SASL enabled, then we need to provide the necessary (no SSL)
```
:thought_balloon: To avoid the potential for duplication here, structuring this block of code like the below pseudo-code would help:
```
if protocol == "SSL" || protocol == "SASL_SSL" {
    getSslElements()
    // set up SSL
}
if protocol == "PLAIN" || protocol == "SASL_PLAIN" {
    getSaslElements()
    // set up SASL
}
```

Does that makes sense?  The idea is that instead of nesting logic, we do SSL setup if we need to, and independently SASL setup if we need to.

agrski replied:
This pattern would then apply to all 3 locations where changes need to happen (currently plaintext is only handled in 2 of them!)

abohmeed replied:
Yes, but if the protocol is `PLAIN` without SASL then why grap SASL username and password at all? afaik `PLAIN` mean no authentication mechanism.

abohmeed replied:
OK after reading your second comment, I will apply the change (if the docs say so 😉)

abohmeed replied:
Done

agrski commented on executor/logger/worker.go:
```diff
@@ -57,7 +57,22 @@ func NewWorker(id int, workQueue chan LogRequest, log logr.Logger, sdepName stri
+				}
+			}
+		}
+		if util.GetKafkaSecurityProtocol() == "SASL_PLAIN" { //if we also have SASL enabled, then we need to provide the necessary (no SSL)
```
:wrench: From checking Kafka docs, I'm pretty sure the security protocol should be `SASL_PLAINTEXT` (or `PLAINTEXT`) instead of `SASL_PLAIN`.
It's the SASL mechanism that might be `PLAIN`.

abohmeed replied:
Done

abohmeed commented:
@agrski I've addressed your latest comments

abohmeed commented:
@RafalSkolasinski @agrski Any updates yet? 🙂

agrski commented:
The docs test failing feels like flakiness given nothing in this PR affects it, so I'll merge in spite of that.
